src/txrequest.h

```diff
@@ -0,0 +1,216 @@
+ * - Whether or not the transaction request failed already (timed out, or NOTFOUND was received).
+ *
+ * Transaction requests are then assigned to peers, following these rules:
+ *
+ * - No transaction is requested as long as another request for the same txhash is outstanding (it needs to fail
+ *   first by passing expiry, or a NOTFOUND or invalid transaction has to be received for it).
+ *
+ *   Rationale: to avoid wasting bandwidth on multiple copies of the same transaction.
+ *
+ * - The same transaction is never requested twice from the same peer, unless the transaction was forgotten in
+ *   between (see next point), and re-announced.
+ *
+ *   Rationale: giving a peer multiple chances to announce a transaction would allow them to bias requests in their
+ *              favor, worsening invblock attacks. The flip side is that as long as an attacker manages to prevent
+ *              us from receiving a transaction, failed announcements (including those from honest peers) will
+ *              linger longer, increasing memory usage somewhat. The impact of this is limited by imposing a cap on
+ *              the number of tracked announcements per peer.
+ *              Invblocking is the practice of announcing transactions but not answer requests for them, in order
+ *              to delay (or prevent) a target learning the transaction. See
+ *              https://allquantor.at/blockchainbib/pdf/miller2015topology.pdf for more information.
+ *
+ * - Announcements are only forgotten about when the peer that announced them went offline, when the transaction
+ *   was received successfully, or when no candidates for a transaction remain that haven't been tried already.
+ *
+ *   Rationale: we need to eventually forget announcements to keep memory bounded, but as long as viable
+ *              candidate peers remain, we prefer to avoid fetching from failed ones. As every request has a finite
+ *              timeout and we schedule new request as soon a previous one expired, there is always progress being
+ *              made towards forgetting a transaction - either successfully or unsuccessfully.
+ *
+ * - Transactions are not requested from a peer until its reqtime has passed.
+ *
+ *   Rationale: enable net_processing code to define a delay for less-than-ideal peers, so that (presumed) better
+ *              peers have a chance to give their announcement first.
+ *
+ * - If multiple viable candidate peers exist according to the above rules, pick a peer as follows:
+ *
+ *   - If any preferred peers are available, non-preferred peers are not considered for what follows.
+ *
+ *     Rationale: preferred peers (outbound, whitelisted) are chosen by us, so are less likely to be under attacker
+ *                control.
+ *
+ *   - Among the remaining candidates, choose the first (non-overloaded) peer to have announced the transaction,
```

Comment:
Okay, so I've run some stats on this now. Results (for me) after just under 230k calls to `RequestedTx` are:

 * in 77.9% of cases, it's requesting the tx from a preferred peer, with no alternative (non-delayed) candidates
 * in 18.2% of cases, it's requesting the tx from a non-preferred peer, with no alternative (n-d) candidates
 * in 1.3% of cases, it's requesting the tx from a preferred peer, with the only (n-d) alternative candidates being non-preferred peers
 * in 1.1% of cases, it's requesting the tx from a preferred peer using the first marker, and other preferred candidates are available (!)
 * in 1.0% of cases, it's requesting the tx from a non-preferred peer using the first marker, with other non-preferred candidates available (and no preferred candidates available) (!)
 * in 0.2% of cases, it's requesting the tx from a non-preferred peer using random selection
 * in 0.02% of cases, it's requesting the tx from a non-preferred peer using random selection

(The remaining ~0.3% I didn't sort through)

This implies that the first marker is only relevant in ~2.1% of cases -- the rest of the time it's determined by the first peer's reqtime being reached at a call to `GetRequestable()` prior to any other announcement for that tx. That occurs because (unless I've missed something) we're calling `GetRequestable()` for each peer about ten times per second (100ms delay in `CConnman::ThreadMessageHandler` between calls to `SendMessages(pnode)`), which means the "first" marker only matters for preferred nodes when the first two preferred nodes to announce a tx manage to both set their reqtimes in the same ~100ms period and thus get activated in the same call the `GetRequestable`.

I think what the "first" marker is actually doing is slightly different to what I was kind-of expecting. ie, "reqtime" and "preferredness" is almost always what's used to actually choose which announcement to request first, and the "first" marker is a backup. It's only relevant when (a) txrequest is used less frequently than every 100ms via some other implementation, (b) the 100ms delay is extended, eg due to cs_main being held while adding a new block or something, or (c) there's a race and we want a consistent tie-breaker.

I haven't instrumented it, but I'm not convinced this does much for dependent transactions. Assuming B spends an output of A, and peers P and Q both announce A prior to B, you could still see announcements as `{Q:A, P:A, P:B}` then call `GetRequestable(P)` and have A assigned to Q and B assigned to P, and requested immediately, with A requested from Q moments later. But if that's not a big problem, what's the advantage in locking  A to Q via the "first" marker?

As it stands, I want to argue that the "first" marker is an unnecessary complication, and maybe we should remove it. But I'm not really convinced -- for one, that claim relies on calling `GetRequestable` very frequently, and maybe we want the flexibility to not do that? But if we called it for a given peer once a second instead of 10x a second, that adds another 0.5s delay on average to requesting a tx, which is pretty significant compared to the existing delays, so maybe relying on calling `GetRequetable` frequently is perfectly reasonable?

Reply:
That's great data, and perhaps not very surprising: we'd hope to pick good peers as outbounds, so most announcements will come in from preferred peers (and even more within the first 2s after the first announcement from inbound connections). As we fetch immediately from preferred connections in general, the "first" marker shouldn't do much.

Given that information, perhaps I wouldn't have added this to the first iteration of this logic. Now that it's implemented, I could argue that it may improve robustness in some cases:
* During the rollout of BIP339, we can expect to see many more transactions fetched from delayed (but possibly preferred) peers, as just one wtxid peer will cause a delay on all announcements from other peers. It's hard to predict/measure what the impact of this will be.
* I can imagine we'd add (possibly random) delays even from fetches from preferred peers, for privacy reasons.
* Very well connected nodes (with way more inbound than outbound nodes) may see a larger percentage fetched from non-preferred connections.

Reply:
I think all those cases would just change the distribution between the cases where the "first" marker doesn't matter -- ie they're entirely determined by reqtime and preferredness.

I do think it might improve robustness *somewhere* though...

Also, if, in the future, we changed the code to either update current_time less, or to target specific reqtimes ("anything between t=0.0 and t=0.499 will be requested at t=1.234"), it might start mattering

> During the rollout of BIP339, we can expect to see many more transactions fetched from delayed (but possibly preferred) peers, as just one wtxid peer will cause a delay on all announcements from other peers.

That's a good point! I've restarted my node since the above numbers, and don't have any logging as to whether any of them were wtxidrelay. However post-restart, I have two inbound wtxidrelay peers, so should already be suffering from BIP339 rollout. And I guess that means all my preferred nodes should be degraded to the same priority as the non-preferred wtxid nodes, which in turn probably explains why I got such a high percentage of non-preferred announcements as the very first request for a txid.

Yeah, collecting `RequestedTx` calls by peerid and whether they were preferred or not gives me:

 *  1049 preferred=0 peer=2717
 *  1500 preferred=0 peer=3993
 *  3255 preferred=0 peer=3815
 *  3270 preferred=1 peer=13
 * 14216 preferred=1 peer=15
 * 14651 preferred=1 peer=14
 * 16918 preferred=1 peer=11
 * 18635 preferred=1 peer=2
 * 30675 preferred=0 peer=1163
 * 37776 preferred=1 peer=1
 * 39962 preferred=1 peer=16
 * 52112 preferred=1 peer=0

(ignoring peers with under 1k requests). Which looks to me like the top 9 peers include the 8 outbounds I connected to initially, and a random inbound wtxidrelay peer which connected about 3h after my peer was restarted. That single node accounts for about 2/3rds of the "18.2% of cases" by the looks.

So I might desable wtxidrelay, and restart my node again, I think, which ironically seems like the easiest way to simulate an all-wtxid-relay world...

Reply:
Ok, updated stats based on 257k `RequestedTx` calls with wtxidrelay disabled (added a `false &&` to the if's in net_processing)

I'm seeing 98.04% of requests being trivial cases -- there's no alternatives that have hit reqtime, and this is the first request of a tx. Despite having lots of (up to ~50?) inbounds, it's still choosing outbounds for almost all txs.

Preferredness only mattered in 0.67% of cases, and only in 0.59% of cases did it matter for the first request.

"First" marker only mattered in 0.77% of cases, and only ever for choosing between two non-preferred peers; though it was at least (almost) always for the first actual request for the txid.

FWIW, ignoring the first hour, I got 520 "accepted orphan tx" messages. I'm running this node with mempool expiry set to 24h so that might be a bit high.

```
86.59% 222556 ABCD  requested  preferred=1  first=1  candidates=[-,-]  completed=[-,-]
11.45%  29429 ABCD  requested  preferred=0  first=1  candidates=[-,-]  completed=[-,-]
 0.18%    455 ABCD  requested  preferred=1  first=0  candidates=[-,-]  completed=[Y,-]
 0.09%    222 ABCD  requested  preferred=0  first=1  candidates=[-,-]  completed=[Y,-]
 0.08%    201 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[-,-]
 0.05%    127 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[Y,Y]
 0.04%    111 ABCD  requested  preferred=1  first=0  candidates=[-,-]  completed=[Y,Y]
 0.03%     79 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[Y,-]
 0.01%     17 ABCD  requested  preferred=1  first=0  candidates=[-,-]  completed=[-,Y]
 0.01%     15 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[-,Y]

 0.59%   1524 ABCD  requested  preferred=1  first=1  candidates=[-,Y]  completed=[-,-]
 0.02%     64 ABCD  requested  preferred=1  first=0  candidates=[Y,Y]  completed=[-,Y]
 0.02%     51 ABCD  requested  preferred=1  first=0  candidates=[Y,Y]  completed=[Y,Y]
 0.02%     50 ABCD  requested  preferred=1  first=0  candidates=[-,Y]  completed=[Y,-]
 0.01%     38 ABCD  requested  preferred=1  first=0  candidates=[-,Y]  completed=[Y,Y]
 0.01%     32 ABCD  requested  preferred=1  first=0  candidates=[Y,-]  completed=[Y,-]
 0.01%     24 ABCD  requested  preferred=1  first=0  candidates=[Y,-]  completed=[Y,Y]
 0.00%     10 ABCD  requested  preferred=1  first=0  candidates=[Y,Y]  completed=[Y,-]
 0.00%      1 ABCD  requested  preferred=1  first=0  candidates=[Y,-]  completed=[-,Y]
 0.00%      1 ABCD  requested  preferred=1  first=0  candidates=[-,Y]  completed=[-,Y]

 0.77%   1978 ABCD  requested  preferred=0  first=1  candidates=[-,Y]  completed=[-,-]
 0.01%     20 ABCD  requested  preferred=0  first=0  candidates=[-,Y]  completed=[Y,Y]
 0.00%      8 ABCD  requested  preferred=0  first=1  candidates=[-,Y]  completed=[Y,-]
 0.00%      2 ABCD  requested  preferred=0  first=0  candidates=[-,Y]  completed=[Y,-]
 0.00%      1 ABCD  requested  preferred=0  first=0  candidates=[-,Y]  completed=[-,Y]
```

Just considering the first request for a tx (ie, `candidates=[-,-] completed=[-,-]` cases) the number of announcements for the tx in CANDIDATE_DELAYED was distributed something like:

```
  48839 ABCD requested preferred=1 delayed=0
  37457 ABCD requested preferred=1 delayed=1
  32913 ABCD requested preferred=1 delayed=2
  27587 ABCD requested preferred=1 delayed=3
  22798 ABCD requested preferred=1 delayed=4
  17525 ABCD requested preferred=1 delayed=5
  12564 ABCD requested preferred=1 delayed=6
   8071 ABCD requested preferred=1 delayed=7
   5148 ABCD requested preferred=1 delayed=8
   2965 ABCD requested preferred=1 delayed=9
   1822 ABCD requested preferred=1 delayed=10

   4437 ABCD requested preferred=0 delayed=0
   2025 ABCD requested preferred=0 delayed=5
   1981 ABCD requested preferred=0 delayed=6
   1780 ABCD requested preferred=0 delayed=4
```

which looks pretty reasonable, I think. (I didn't break the delayed count into preferred/non-preferred, so the delayed= figure should include both inbounds and any overloaded outbounds)

I wasn't quite as lucky with my outbounds staying around the entire timeby the looks, but got a much more even looking distribution of requests between the different prefrences:

```
   1048 ABCD requested preferred=0 peer=2013
   1294 ABCD requested preferred=0 peer=4076
   1359 ABCD requested preferred=0 peer=4096
   1725 ABCD requested preferred=0 peer=235
   2193 ABCD requested preferred=0 peer=3370
   2630 ABCD requested preferred=0 peer=1950
   2694 ABCD requested preferred=0 peer=2304
   2843 ABCD requested preferred=0 peer=2095
   3618 ABCD requested preferred=0 peer=1853

      9 ABCD requested preferred=1 peer=6529
     21 ABCD requested preferred=1 peer=5853
     25 ABCD requested preferred=1 peer=1737
     39 ABCD requested preferred=1 peer=1624
     44 ABCD requested preferred=1 peer=30
     57 ABCD requested preferred=1 peer=644
    853 ABCD requested preferred=1 peer=22
   2312 ABCD requested preferred=1 peer=11
   5470 ABCD requested preferred=1 peer=202
  13974 ABCD requested preferred=1 peer=0
  14062 ABCD requested preferred=1 peer=2238
  14427 ABCD requested preferred=1 peer=14
  16520 ABCD requested preferred=1 peer=16
  23714 ABCD requested preferred=1 peer=3385
  37178 ABCD requested preferred=1 peer=24
  46702 ABCD requested preferred=1 peer=29
  49527 ABCD requested preferred=1 peer=44
```

(ignoring non-preferred peers that didn't have >1000 requests)

Reply:
Am currently running a test to see if removing "first" markers has a noticable effect on orphans (or anything else), but my current feeling is that the benefits of "first" markers don't justify the complexity. I've had a quick go at pulling the feature out into its own commit at https://github.com/ajtowns/bitcoin/commits/202009-txoverhaul-sep-first to see how much complexity that is, fwiw.

I think the robustness that it adds is preserving the "first announcer" when multiple candidate announcements become ready at the same time. At present it only does that when their reqtimes were very similar from the word go, relative to how often `GetRequestable` is called.

One thing that might be worth considering is extending that towards a firmer guarantee, eg if a later, equally-preferred peer's announcement ends up with an earlier reqtime (due to time going backwards or randomized delays maybe) automatically adjust the "first" peer's announcement's reqtime to the same figure. (That would make a difference even with the current net_processing code in this PR, I think. Namely in the case where a non-wtxid-relay announces txid X, then a wtxid-relay-node announces wtxid X, ie X has no witness -- it'd bump the reqtime of the original peer's announcement at that point, rather than making it wait for the full TXID_RELAY_DELAY)

Even if that's a good idea though, I'm not really convinced it makes sense to do it now, rather than later, with code to actually randomize times and thus make it relevant, and we can see if it covers all the cases that actually make sense. eg, would probably want to bump the reqtime of the "first" preferred announcement upon receipt of a non-preferred announcement with earlier reqtime, but probably not the reverse...

Reply:
Done, I've changed this PR to drop the "first" marker logic, based on your branch + adjusting fuzz tester and some extra comments. The commit to add it back is here: https://github.com/sipa/bitcoin/commits/202009_txrequest_rand_wtxid_first

Reply:
(replying to resolved thread without marking it unresolved)

With "first" marker disabled (essentially treating every peer as oveloaded), I got:

``` 324484 ABCD  requested  preferred=1  first=0  candidates=[-,-]  completed=[-,-]  -- 81.7%
  61457 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[-,-] -- 15.5%
    599 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[Y,Y]
    122 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[-,Y]
    110 ABCD  requested  preferred=0  first=0  candidates=[-,-]  completed=[Y,-]
     87 ABCD  requested  preferred=1  first=0  candidates=[-,-]  completed=[Y,Y]
     60 ABCD  requested  preferred=1  first=0  candidates=[-,-]  completed=[Y,-]
     13 ABCD  requested  preferred=1  first=0  candidates=[-,-]  completed=[-,Y]

   2949 ABCD  requested  preferred=1  first=0  candidates=[-,Y]  completed=[-,-] -- 0.7%
    543 ABCD  requested  preferred=1  first=0  candidates=[-,Y]  completed=[Y,Y]
    214 ABCD  requested  preferred=1  first=0  candidates=[Y,Y]  completed=[Y,Y]
     51 ABCD  requested  preferred=1  first=0  candidates=[Y,Y]  completed=[-,Y]
     24 ABCD  requested  preferred=1  first=0  candidates=[-,Y]  completed=[-,Y]
      5 ABCD  requested  preferred=1  first=0  candidates=[Y,Y]  completed=[Y,-]
      5 ABCD  requested  preferred=1  first=0  candidates=[-,Y]  completed=[Y,-]
      4 ABCD  requested  preferred=1  first=0  candidates=[Y,-]  completed=[Y,Y]
      3 ABCD  requested  preferred=1  first=0  candidates=[Y,-]  completed=[Y,-]
      1 ABCD  requested  preferred=1  first=0  candidates=[Y,-]  completed=[-,Y]

   3285 ABCD  requested  preferred=0  first=0  candidates=[-,Y]  completed=[-,-] -- 0.8%
   2910 ABCD  requested  preferred=0  first=0  candidates=[-,Y]  completed=[Y,Y] -- 0.7%
    107 ABCD  requested  preferred=0  first=0  candidates=[-,Y]  completed=[-,Y]
      8 ABCD  requested  preferred=0  first=0  candidates=[-,Y]  completed=[Y,-]
```

I got 561 "accepted orphan tx" entries (ignoring the first first hour), compared to 520 last time, which is an 8% increase in orphans compared to running for about 90% longer or doing 54% more requests; so seems like any differences in behaviour are lost in the noise.

Looking at the per-peer behaviour, one non-preferred peer got a lot of requests:

```
   1088 ABCD requested preferred=0 peer=8236
   1245 ABCD requested preferred=0 peer=7422
   1470 ABCD requested preferred=0 peer=7219
   1486 ABCD requested preferred=0 peer=7712
   1540 ABCD requested preferred=0 peer=7342
  13124 ABCD requested preferred=0 peer=53
  22367 ABCD requested preferred=0 peer=12516

    159 ABCD requested preferred=1 peer=20
    166 ABCD requested preferred=1 peer=7
    746 ABCD requested preferred=1 peer=6
  16624 ABCD requested preferred=1 peer=31
  22917 ABCD requested preferred=1 peer=122
  25830 ABCD requested preferred=1 peer=24
  54666 ABCD requested preferred=1 peer=30
  55279 ABCD requested preferred=1 peer=27
  72306 ABCD requested preferred=1 peer=29
  79637 ABCD requested preferred=1 peer=23
```

(ignores preferred peers that got less than 100 requests, and non-preferred peers that got less than 1000).

For peer=12516, almost every request (21625) was when it was the only candidate and no other peer had already been tried for the tx. It looks like it just happened to be fastest most of the time, with plenty of other inbounds (presumably) having already announced when we actually make the request:

```
...
      7 delayed=60
     12 delayed=59
     12 delayed=58
...
    744 delayed=12
    904 delayed=11
   1025 delayed=10
   1146 delayed=9
   1252 delayed=8
   1389 delayed=7
   1448 delayed=6
   1335 delayed=5
   1354 delayed=4
   1096 delayed=3
    706 delayed=2
    420 delayed=1
    193 delayed=0
```

Results for peer=53 look pretty similar to me, though it had a higher proportion of announcements where delayed=0 or delayed=1.

So I think that just means it's showing a heavy bias towards fast peers/first announcers? The bias part is fine and desirable, but maybe it would be good to reduce the heaviness in future by mildly varying peers' delays, either randomly or based on load in some way? Something to worry about in a followup.